flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SortMergeResultPartitionReadScheduler.java

```diff
@@ -0,0 +1,365 @@
+
+    /** Lock used to synchronize multi-thread access to thread-unsafe fields. */
+    private final Object lock;
+
+    /** Corresponding {@link SortMergeResultPartition} to read data from. */
+    private final SortMergeResultPartition partition;
+
+    /** Buffer pool from which to allocate buffers for shuffle data reading. */
+    private final BatchShuffleReadBufferPool bufferPool;
+
+    /** Executor to run the shuffle data reading task. */
+    private final Executor ioExecutor;
+
+    /** Maximum number of buffers can be allocated by this partition reader. */
+    private final int maxRequestedBuffers;
+
+    /** All failed subpartition readers to be released. */
+    @GuardedBy("lock")
+    private final Set<SortMergeSubpartitionReader> failedReaders = new HashSet<>();
+
+    /** All readers waiting to read data of different subpartitions. */
+    @GuardedBy("lock")
+    private final Set<SortMergeSubpartitionReader> allReaders = new HashSet<>();
+
+    /** File channel shared by all subpartitions to read data from. */
+    @GuardedBy("lock")
+    private FileChannel dataFileChannel;
+
+    /** File channel shared by all subpartitions to read index from. */
+    @GuardedBy("lock")
+    private FileChannel indexFileChannel;
+
+    /** Number of buffers already allocated and still not recycled by this partition reader. */
+    @GuardedBy("lock")
+    private int numRequestedBuffers;
+
+    /**
+     * Whether the data reading task is currently running or not. This flag is used when trying to
+     * submit the data reading task.
+     */
+    @GuardedBy("lock")
+    private boolean isRunning;
+
+    /** Whether this reader has been released or not. */
+    @GuardedBy("lock")
+    private boolean isReleased;
+
+    public SortMergeResultPartitionReadScheduler(
+            SortMergeResultPartition partition,
+            BatchShuffleReadBufferPool bufferPool,
+            Executor ioExecutor,
+            Object lock) {
+        this.lock = checkNotNull(lock);
+        this.bufferPool = checkNotNull(bufferPool);
+        this.ioExecutor = checkNotNull(ioExecutor);
+        this.partition = checkNotNull(partition);
+        // one partition reader can consume at most 32M (the expected buffers per request is 8M)
+        // buffers for data read. Currently, it is only an empirical value can not be configured
+        this.maxRequestedBuffers = Math.max(1, 4 * bufferPool.getNumBuffersPerRequest());
+
+        // initialize the buffer pool eagerly to avoid reporting errors like OOM too late
+        bufferPool.initialize();
+    }
+
+    @Override
+    public synchronized void run() {
+        Set<SortMergeSubpartitionReader> finishedReaders = new HashSet<>();
+        Queue<SortMergeSubpartitionReader> availableReaders = getAvailableReaders();
+
+        Queue<MemorySegment> buffers = new ArrayDeque<>();
+        if (!availableReaders.isEmpty()) {
+            try {
+                buffers.addAll(bufferPool.requestBuffers());
```

Comment:
Seems the scheduler can be blocked here even if the scheduler is released when waiting for buffers.

Reply:
When released, fail all readers immediately to trigger task failure of downstream is better.

Reply:
The block of the thread hurt nothing, if there is enough buffers, the method should return immediately and the requested buffer will be recycled soon, if there is no enough buffers, blocking request does not have a side effect.

Reply:
Looks to me this thread can be blocked for 5min if there is not enough buffers, even if the partition and task has been disposed. This is undesirable I think.

Reply:
I will move the timeout logic out of the buffer pool to the read scheduler then we can solve this problem.